### PR TITLE
 Fix issue 20019: Symbol not found: _dyld_enumerate_tlv_storage on macOS 10.15

### PR DESCRIPTION
**Pull request by jacob-carlborg**
The symbol `dyld_enumerate_tlv_storage` has been used to iterate all the thread local storage addresses and registered them as roots with the garbage collector.

The symbol has always been a private symbol, not part of the public API. In the latest version of macOS, 10.15 it has been removed.

This replaces the usage of `dyld_enumerate_tlv_storage` by manually iterating the TLV sections of all images and get key that the dynamic loader has setup for each TLS symbol. The key is then used to get the base address of the TLV storage.

The advantage of this change is that it doesn't rely on any private symbols anymore. The disadvantage is that it relies on some implementation details of the dynamic loader. So this might be even worse than the usage of the private symbol `dyld_enumerate_tlv_storage`.

This fix is backwards compatible and works with all version of macOS that D supports.

This PR depends on https://github.com/dlang/druntime/pull/2665.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
Should I target stable? This depends on https://github.com/dlang/druntime/pull/2665.

**Comment by Geod24** [comment]
If I understand the implications correctly, it means that not a single D program will work on OSX 10.15 until they are compiled with a release which includes this PR ?

**Comment by jacob-carlborg** [comment]
> If I understand the implications correctly, it means that not a single D program will work on OSX 10.15 until they are compiled with a release which includes this PR ?

That's correct.

**Comment by thewilsonator** [comment]
> Should I target stable?

Yes.

**Comment by jacob-carlborg** [comment]
Targeting stable now.

**Comment by p0nce** [comment]
Will programs compiled with LDC also need this fix?

**Comment by thewilsonator** [comment]
Yes..

File: src/rt/sections_osx_x86_64.d [hunk 1]
```diff
@@ -183,3 +178,177 @@ ubyte[] getSection(in mach_header* header, intptr_t slide,
...
+extern (C) size_t malloc_size(const void* ptr) nothrow @nogc;
+
+/// Represents a TLS range.
+struct TLSRange
+{
+    /// The start of the range.
+    void* start;
+
+    /// The size of the range.
+    size_t size;
```
Comment (WalterBright) [inline src/rt/sections_osx_x86_64.d]: The fields should be reversed, so they match D slices.

File: src/rt/sections_osx_x86_64.d [hunk 2]
```diff
@@ -183,3 +178,177 @@ ubyte[] getSection(in mach_header* header, intptr_t slide,
...
+    {
+        return start !is null && size > 0;
+    }
+
+    /// Returns the range as an array.
+    void[] toArray() pure nothrow @nogc
+    {
+        return start[0 .. size];
+    }
+}
```
Comment (WalterBright) [inline src/rt/sections_osx_x86_64.d]: I can't see the difference between this struct and a simple D dynamic array.

**Comment by jacob-carlborg** [comment]
FYI, I have now finally got a reply from Apple regarding this issue. I asked if there's an alternative to the `_dyld_enumerate_tlv_storage` symbol. The answer was basically what I have already implemented here.

**Pull request merged**